lib/cadet_web/controllers/generate_ai_comments.ex

```diff
@@ -0,0 +1,424 @@
+
+    summary("Generate AI comments for a given submission.")
+
+    security([%{JWT: []}])
+
+    consumes("application/json")
+    produces("application/json")
+
+    parameters do
+      courseId(:path, :integer, "course id", required: true)
+      submissionId(:path, :integer, "submission id", required: true)
+      questionId(:path, :integer, "question id", required: true)
+    end
+
+    response(200, "OK", Schema.ref(:GenerateAIComments))
+    response(400, "Invalid or missing parameter(s) or submission and/or question not found")
+    response(401, "Unauthorized")
+    response(403, "Forbidden")
+    response(403, "LLM grading is not enabled for this course")
+  end
+
+  swagger_path :save_final_comment do
+    post("/courses/{courseId}/admin/save-final-comment/{submissionId}/{questionId}")
+
+    summary("Save the final comment chosen for a submission.")
+
+    security([%{JWT: []}])
+
+    consumes("application/json")
+    produces("application/json")
+
+    parameters do
+      submissionId(:path, :integer, "submission id", required: true)
+      questionId(:path, :integer, "question id", required: true)
+      comment(:body, :string, "The final comment to save", required: true)
+    end
+
+    response(200, "OK", Schema.ref(:SaveFinalComment))
+    response(400, "Invalid or missing parameter(s)")
+    response(401, "Unauthorized")
+    response(403, "Forbidden")
+  end
+
+  def swagger_definitions do
+    %{
+      GenerateAIComments:
+        swagger_schema do
+          properties do
+            comments(:string, "AI-generated comments on the submission answers")
+          end
+        end,
+      SaveFinalComment:
+        swagger_schema do
+          properties do
+            status(:string, "Status of the operation")
+          end
+        end
+    }
+  end
+
+  defp decrypt_llm_api_key(nil), do: nil
+
+  defp decrypt_llm_api_key(encrypted_key) do
+    case Application.get_env(:openai, :encryption_key) do
+      secret when is_binary(secret) and byte_size(secret) >= 16 ->
+        key = binary_part(secret, 0, min(32, byte_size(secret)))
+
+        case Base.decode64(encrypted_key) do
+          {:ok, decoded} ->
+            iv = binary_part(decoded, 0, 16)
+            tag = binary_part(decoded, 16, 16)
+            ciphertext = binary_part(decoded, 32, byte_size(decoded) - 32)
+
+            case :crypto.crypto_one_time_aead(:aes_gcm, key, iv, ciphertext, "", tag, false) do
+              plain_text when is_binary(plain_text) -> {:ok, plain_text}
+              _ -> {:decrypt_error, :decryption_failed}
+            end
+
+          _ ->
+            Logger.error(
+              "Failed to decode encrypted key, is it a valid AES-256 key of 16, 24 or 32 bytes?"
+            )
+
+            {:decrypt_error, :decryption_failed}
+        end
+
+      _ ->
+        Logger.error("Encryption key not configured")
+        {:decrypt_error, :invalid_encryption_key}
+    end
+  end
+end
```

Comment:
Should be abstracted away together with the encrypt function in a separate module for better abstraction/encapsulation